src/content/components/TimelineCanvas.js

```diff
@@ -0,0 +1,176 @@
+// @flow
+import React, { Component } from 'react';
+import { timeCode } from '../../common/time-code';
+import classNames from 'classnames';
+
+import type { CssPixels, DevicePixels, NonNull } from '../../common/types/units';
+
+type HoveredItem = NonNull;
+
+type Props = {
+  containerWidth: CssPixels,
+  containerHeight: CssPixels,
+  className: string,
+  onDoubleClickItem: HoveredItem => void,
+  getHoveredItemInfo: HoveredItem => string,
+  drawCanvas: (CanvasRenderingContext2D, HoveredItem) => void,
+  hitTest: (x: CssPixels, y: CssPixels) => null | HoveredItem,
+};
+
+require('./TimelineCanvas.css');
+
+export default class TimelineCanvas extends Component {
+
+  props: Props
+  _requestedAnimationFrame: boolean
+  _devicePixelRatio: 1
+  _ctx: CanvasRenderingContext2D
+  state: {
+    hoveredItem: null | HoveredItem;
+  }
+
+  constructor(props: Props) {
+    super(props);
+    this._requestedAnimationFrame = false;
+    this._devicePixelRatio = 1;
+    this.state = { hoveredItem: null };
+
+    (this: any).onMouseMove = this.onMouseMove.bind(this);
+    (this: any).onMouseOut = this.onMouseOut.bind(this);
+    (this: any).onDoubleClick = this.onDoubleClick.bind(this);
+    (this: any).getHoveredItemInfo = this.getHoveredItemInfo.bind(this);
+  }
+
+  shouldComponentUpdate() {
+    // If the parent updates, always re-render.
+    return true;
+  }
+
+  _scheduleDraw() {
+    const { className, drawCanvas } = this.props;
+    if (!this._requestedAnimationFrame) {
+      this._requestedAnimationFrame = true;
+      window.requestAnimationFrame(() => {
+        this._requestedAnimationFrame = false;
+        if (this.refs.canvas) {
+          timeCode(`${className} render`, () => {
+            this._prepCanvas();
+            drawCanvas(this._ctx, this.state.hoveredItem);
+          });
+        }
+      });
+    }
+  }
+
+  _prepCanvas() {
+    const {canvas} = this.refs;
+    const {containerWidth, containerHeight} = this.props;
+    const {devicePixelRatio} = window;
+    const pixelWidth: DevicePixels = containerWidth * devicePixelRatio;
+    const pixelHeight: DevicePixels = containerHeight * devicePixelRatio;
+    if (!this._ctx) {
+      this._ctx = canvas.getContext('2d');
+    }
```

Comment:
I believe this pattern was to satisfy Flow, as the ctx can be null. I agree that it's odd. I'll add a comment about the flow requirement to clear it up.